File: Python/ShortestUnsortedContinuousSubarray/ShortestUnsortedContinuousSubarray.py
class Solution(object):
    # (95ms)
    def findUnsortedSubarray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        left, right, sortedNums = 0, len(nums) - 1, sorted(nums[:])
        while left < len(nums):
            if nums[left] != sortedNums[left]:
                break
            left += 1
        while right >= 0:
            if nums[right] != sortedNums[right]:
                break
            right -= 1
        return right - left + 1 if left < right else 0

    # (149ms)
    def findUnsortedSubarray2(self, nums):
        begin, end, minNum, maxNum = -1, -2, nums[-1], nums[0]
        for i in xrange(1, len(nums)):
            maxNum, minNum = max(maxNum, nums[i]), min(minNum, nums[-i - 1])
            if maxNum > nums[i]:
                end = i
            if minNum < nums[-i - 1]:
                begin = len(nums) - i - 1
        return end - begin + 1

File: JavaScript/ShortestUnsortedContinuousSubarray/ShortestUnsortedContinuousSubarray.js
/**
 * (148ms)
 * @param {number[]} nums
 * @return {number}
 */
var findUnsortedSubarray = function(nums) {
    var begin = -1, end = -2, n = nums.length;
    var min = nums[n - 1], max = nums[0];
    for(var i = 1; i < n; i++) {
    	max = Math.max(max, nums[i]);
    	min = Math.min(min, nums[n - 1 - i]);
    	if(max > nums[i])
    		end = i;
    	if(min < nums[n - 1 - i]) 
    		begin = n - 1 - i;
    }
    return end - begin + 1;
};